plugins/hosts/darwin/cap/fs_iso.rb

```diff
@@ -0,0 +1,78 @@
+require "tempfile"
+require 'fileutils'
+require 'pathname'
+require "vagrant/util/subprocess"
+require "vagrant/util/map_command_options"
+require "vagrant/util/directory"
+
+module VagrantPlugins
+  module HostDarwin
+    module Cap
+      class FsISO
+        @@logger = Log4r::Logger.new("vagrant::host::darwin::fs_iso")
+
+        BUILD_ISO_CMD = "hdiutil".freeze
+
+        # Check that the host has the ability to generate ISOs
+        #
+        # @param [Vagrant::Environment] env
+        # @return [Boolean]
+        def self.isofs_available(env)
+          !!Vagrant::Util::Which.which(BUILD_ISO_CMD)
+        end
+
+        # Generate an ISO file of the given source directory
+        #
+        # @param [Vagrant::Environment] env
+        # @param [String] source_directory Contents of ISO
+        # @param [Map] extra arguments to pass to the iso building command
+        #              :file_destination (string) location to store ISO
+        #              :volume_id (String) to set the volume name 
+        # @return [Pathname] ISO location
+        # @note If file_destination exists, source_directory will be checked
+        #       for recent modifications and a new ISO will be generated if requried.
+        def self.create_iso(env, source_directory, **extra_opts)
+          file_destination = extra_opts[:file_destination]
+          source_directory = Pathname.new(source_directory)
+          if file_destination.nil?
+            @@logger.info("No file destination specified, creating temp location")
+            tmpfile = Tempfile.new(["vagrant", ".iso"])
+            file_destination = Pathname.new(tmpfile.path)
+            tmpfile.delete
+          else
+            file_destination = Pathname.new(file_destination.to_s)
+            # If the file destination path is a folder, target the output to a randomly named
+            # file in that dir
+            if file_destination.extname != ".iso"
+              file_destination = file_destination.join("#{rand(36**6).to_s(36)}_vagrant.iso")
+            end
+          end
+          # Ensure destination directory is available
+          FileUtils.mkdir_p(File.dirname(file_destination.to_s))
```

Comment:
How do you feel about changing this to `FileUtils.mkdir_p(file_destination.dirname)`. Using the `file_destination.dirname.mkpath` requires building out far more extensive mocks in the tests. I think it makes the tests way harder to read and debug. For example, in order to test this function with `file_destination.dirname.mkpath` I tried to do something like:

```
let(:file_destination) { "/woo/out.iso" }
let(:file_destination_path) {Pathname.new( "/woo/out.iso" )}

before do
  allow(Pathname).to receive(:new).with(file_destination).and_return(file_destination_path) 
  # make sure no files are actually created for test
  allow(file_destination_path).to receive_message_chain(:dirname, :mkpath)
end
```

maybe my approach to testing this is wrong though?

Reply:
I think you could do something like this:

```ruby
require "tmpdir"

let(:file_destination) { @fd ||= Pathname.new(Dir.mktmpdir).join("out.iso").to_s }
let(:file_destination_path) { Pathname.new(file_destination) }

before { allow(Pathname).to receive(:new).with(file_destination).and_return(file_destination_path) }
after { file_destination_path.dirname.delete }
```

which would allow you to have a proper pathname instance that wouldn't need anything mocked, but you could include checks that expected behaviors were occurring, like:

```
expect(file_destination_path).to receive_message_chain(:dirname, :mkpath)
```

However, keeping it with the `#mkdir_p` is fine as they are functionally equivalent. Within the tests, though, we should be checking the the parameter provided to the mock (`expect(FileUtils).to receive(:mkdir_p).with("expected/prefix")`) so it's only matching what we expect to be matched.